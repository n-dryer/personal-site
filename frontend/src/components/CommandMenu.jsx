import React, { useState, useEffect, useRef } from 'react';
import { Command } from 'cmdk';
import { Download, Mail, Linkedin, Github, Moon, Sun, ChevronRight } from 'lucide-react';

const CommandMenu = ({ userData, toggleTheme, darkMode }) => {
  const [open, setOpen] = useState(false);
  const [inputValue, setInputValue] = useState('');
  const inputRef = useRef(null);
  const listRef = useRef(null);

  // Define command items
  const items = [
    { 
      label: "Download résumé", 
      icon: Download, 
      href: "/nathan_dryer_resume.pdf", 
      shortcut: "⏎", 
      primary: true 
    },
    { 
      label: "Email Nate", 
      icon: Mail, 
      href: `mailto:${userData.email}` 
    },
    { 
      label: "LinkedIn", 
      icon: Linkedin, 
      href: userData.socialLinks.find(link => link.name === "LinkedIn")?.url || "#" 
    },
    { 
      label: "GitHub", 
      icon: Github, 
      href: userData.socialLinks.find(link => link.name === "GitHub")?.url || "#" 
    },
    { 
      label: `Toggle ${darkMode ? 'light' : 'dark'} mode`, 
      icon: darkMode ? Sun : Moon, 
      action: "theme" 
    },
  ];

  // Filter items based on search input
  const filteredItems = inputValue === '' 
    ? items 
    : items.filter(item => 
        item.label.toLowerCase().includes(inputValue.toLowerCase())
      );

  // Handle keyboard shortcuts
  useEffect(() => {
    const handleKeyDown = (e) => {
      // Open command palette with Cmd+K or Ctrl+K
      if ((e.metaKey || e.ctrlKey) && e.key === 'k') {
        e.preventDefault();
        setOpen(true);
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => window.removeEventListener('keydown', handleKeyDown);
  }, []);

  // Auto-focus input on palette open only on desktop
  useEffect(() => {
    if (open && window.innerWidth >= 768) {
      setTimeout(() => {
        inputRef.current?.focus();
      }, 100);
    }
  }, [open]);

  // Handle email copy
  const handleEmailCopy = () => {
    navigator.clipboard.writeText(userData.email);
    // You could add a toast notification here
    setOpen(false);
  };

  return (
    <Command
      label="Site command palette"
      className="w-full max-w-full md:max-w-md mx-auto rounded-2xl bg-white/65 dark:bg-slate-900/50
                backdrop-blur-sm shadow-lg ring-1 ring-black/5 dark:ring-white/10
                focus-within:ring-blue-500 dark:focus-within:ring-blue-400 transition"
    >
      <Command.Input
        ref={inputRef}
        value={inputValue}
        onValueChange={setInputValue}
        placeholder="Type a command…"
        className="h-12 w-full bg-transparent px-4 text-base text-gray-800 dark:text-white outline-none"
        autoComplete="off"
      />
      <Command.List 
        ref={listRef}
        className="grid gap-1 p-2"
      >
        {filteredItems.map(({ label, icon: Icon, href, shortcut, primary, action }) => (
          <Command.Item 
            key={label} 
            className="rounded-xl focus:bg-slate-100 dark:focus:bg-white/5 outline-none"
            role="menuitem"
          >
            {href ? (
              <a
                href={href}
                className={`flex items-center gap-3 px-4 py-3 rounded-xl w-full min-h-[48px]
                          ${primary 
                            ? "bg-gradient-to-r from-blue-500 to-blue-600 text-white hover:from-blue-600 hover:to-blue-700" 
                            : "hover:bg-slate-100 dark:hover:bg-white/5"}`}
                download={href.includes('resume') ? "Nathan_Dryer_Resume.pdf" : undefined}
              >
                <div className={`flex items-center justify-center w-8 h-8 ${primary 
                  ? "bg-white/20 rounded-full" 
                  : "text-blue-500 dark:text-blue-400"}`}>
                  <Icon className="w-5 h-5 shrink-0" />
                </div>
                <span className="grow text-left">{label}</span>
                {shortcut && (
                  <kbd className="px-2 py-1 text-xs bg-black/5 dark:bg-white/10 rounded">
                    {shortcut}
                  </kbd>
                )}
                {!primary && !shortcut && <ChevronRight className="w-4 h-4 opacity-50" />}
              </a>
            ) : (
              <button
                onClick={() => {
                  if (action === "theme") {
                    toggleTheme();
                  }
                  setOpen(false);
                }}
                className="flex items-center gap-3 px-4 py-3 rounded-xl w-full min-h-[48px]
                          hover:bg-slate-100 dark:hover:bg-white/5"
              >
                <div className="flex items-center justify-center w-8 h-8 text-blue-500 dark:text-blue-400">
                  <Icon className="w-5 h-5 shrink-0" />
                </div>
                <span className="grow text-left text-gray-800 dark:text-white">{label}</span>
              </button>
            )}
          </Command.Item>
        ))}
      </Command.List>
    </Command>
  );
};

export default CommandMenu;
