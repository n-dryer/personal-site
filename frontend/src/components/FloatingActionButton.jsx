import React, { useState, useEffect, useRef } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import CopyToClipboard from 'react-copy-to-clipboard';

const FloatingActionButton = ({ email }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [showToast, setShowToast] = useState(false);
  const [toastMessage, setToastMessage] = useState('');
  const containerRef = useRef(null);

  useEffect(() => {
    const handleClickOutside = (e) => {
      if (containerRef.current && !containerRef.current.contains(e.target)) {
        setIsOpen(false);
      }
    };

    if (isOpen) {
      document.addEventListener('mousedown', handleClickOutside);
    }
    
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [isOpen]);

  const handleCopy = () => {
    setToastMessage('Email copied to clipboard!');
    setShowToast(true);
    setTimeout(() => setShowToast(false), 2000);
  };

  const scrollToTop = () => {
    document.getElementById('top')?.scrollIntoView({ behavior: 'smooth' });
  };

  const scrollToTimeline = () => {
    document.getElementById('timeline')?.scrollIntoView({ behavior: 'smooth' });
  };

  const scrollToSkills = () => {
    document.getElementById('skills')?.scrollIntoView({ behavior: 'smooth' });
  };

  const scrollToBottom = () => {
    document.getElementById('end')?.scrollIntoView({ behavior: 'smooth' });
  };

  const downloadResume = () => {
    const link = document.createElement('a');
    link.href = '/nathan_dryer_resume.pdf';
    link.download = 'Nathan_Dryer_Resume.pdf';
    link.click();
    
    setToastMessage('Resume downloaded!');
    setShowToast(true);
    setTimeout(() => setShowToast(false), 2000);
  };

  // Button configuration
  const buttons = [
    { label: "Top", icon: "arrow-up", onClick: scrollToTop },
    { label: "Timeline", icon: "timeline", onClick: scrollToTimeline },
    { label: "Skills", icon: "lightbulb", onClick: scrollToSkills },
    { label: "Contact", icon: "chevron-down", onClick: scrollToBottom },
    { 
      label: "Email", 
      icon: "mail",
      onClick: handleCopy,
      copyToClipboard: true,
      text: email
    },
    { 
      label: "Resume", 
      icon: "document-download",
      onClick: downloadResume 
    }
  ];

  // Circular menu layout calculations
  const getButtonPosition = (index, total, radius = 80) => {
    // Start from the top position and distribute buttons evenly in a circle
    const angleOffset = -90; // Start from top (negative y-axis in screen coordinates)
    const angleStep = (360 / total);
    const angle = ((index * angleStep) + angleOffset) * (Math.PI / 180);
    
    return {
      x: Math.cos(angle) * radius,
      y: Math.sin(angle) * radius
    };
  };

  // SVG path map
  const iconPathMap = {
    "arrow-up": "M5 15l7-7 7 7",
    "timeline": "M4 5h16M4 12h16M4 19h7",
    "lightbulb": "M9.663 17h4.673M12 3v1m6.364 1.636l-.707.707M21 12h-1M4 12H3m3.343-5.657l-.707-.707m2.828 9.9a5 5 0 117.072 0l-.548.547A3.374 3.374 0 0014 18.469V19a2 2 0 11-4 0v-.531c0-.895-.356-1.754-.988-2.386l-.548-.547z",
    "chevron-down": "M19 9l-7 7-7-7",
    "mail": "M3 8l7.89 5.26a2 2 0 002.22 0L21 8M5 19h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z",
    "document-download": "M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
  };

  return (
    <>
      <div ref={containerRef} className="fixed bottom-8 right-8 z-50">
        <div className="relative">
          <AnimatePresence>
            {isOpen && (
              <div className="absolute top-0 left-0">
                {buttons.map((button, index) => {
                  const position = getButtonPosition(index, buttons.length);
                  
                  if (button.copyToClipboard) {
                    return (
                      <CopyToClipboard key={index} text={button.text} onCopy={button.onClick}>
                        <motion.button
                          className="menu-button"
                          initial={{ scale: 0, x: 0, y: 0, opacity: 0 }}
                          animate={{ 
                            scale: 1, 
                            x: position.x, 
                            y: position.y,
                            opacity: 1,
                          }}
                          exit={{ 
                            scale: 0,
                            x: 0,
                            y: 0,
                            opacity: 0,
                            transition: { duration: 0.2 }
                          }}
                          transition={{ 
                            type: "spring",
                            stiffness: 350,
                            damping: 25,
                            delay: index * 0.04,
                          }}
                          aria-label={button.label}
                          whileHover={{ 
                            scale: 1.1, 
                            backgroundColor: "#4F46E5",
                            color: "white"
                          }}
                          whileTap={{ scale: 0.95 }}
                        >
                          <svg xmlns="http://www.w3.org/2000/svg" className="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d={iconPathMap[button.icon]} />
                          </svg>
                        </motion.button>
                      </CopyToClipboard>
                    );
                  }

                  return (
                    <motion.button
                      key={index}
                      className="menu-button"
                      initial={{ scale: 0, x: 0, y: 0, opacity: 0 }}
                      animate={{ 
                        scale: 1, 
                        x: position.x, 
                        y: position.y,
                        opacity: 1,
                      }}
                      exit={{ 
                        scale: 0,
                        x: 0,
                        y: 0,
                        opacity: 0,
                        transition: { duration: 0.2 }
                      }}
                      transition={{ 
                        type: "spring",
                        stiffness: 350,
                        damping: 25,
                        delay: index * 0.04,
                      }}
                      onClick={button.onClick}
                      aria-label={button.label}
                      whileHover={{ 
                        scale: 1.1, 
                        backgroundColor: "#4F46E5",
                        color: "white"
                      }}
                      whileTap={{ scale: 0.95 }}
                    >
                      <svg xmlns="http://www.w3.org/2000/svg" className="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                        <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d={iconPathMap[button.icon]} />
                      </svg>
                    </motion.button>
                  );
                })}
              </div>
            )}
          </AnimatePresence>

          <motion.button
            className="main-button"
            onClick={() => setIsOpen(!isOpen)}
            whileHover={{ scale: 1.1 }}
            whileTap={{ scale: 0.95 }}
            animate={isOpen ? { rotate: 135 } : { rotate: 0 }}
            transition={{ type: "spring", stiffness: 500, damping: 30 }}
            aria-label={isOpen ? "Close menu" : "Open menu"}
            aria-expanded={isOpen}
          >
            <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 6v6m0 0v6m0-6h6m-6 0H6" />
            </svg>
          </motion.button>
        </div>
      </div>

      <AnimatePresence>
        {showToast && (
          <motion.div
            className="toast"
            initial={{ opacity: 0, y: 50 }}
            animate={{ opacity: 1, y: 0 }}
            exit={{ opacity: 0 }}
            transition={{ duration: 0.3 }}
            role="alert"
          >
            {toastMessage}
          </motion.div>
        )}
      </AnimatePresence>
    </>
  );
};

export default FloatingActionButton;
