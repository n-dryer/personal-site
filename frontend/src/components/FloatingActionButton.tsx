import React, { useState } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import CopyToClipboard from 'react-copy-to-clipboard';

interface FABProps {
  email: string;
}

const FloatingActionButton: React.FC<FABProps> = ({ email }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [showToast, setShowToast] = useState(false);

  const handleCopy = () => {
    setShowToast(true);
    setIsOpen(false);
    setTimeout(() => setShowToast(false), 2000);
  };

  const scrollToTop = () => {
    document.getElementById('top')?.scrollIntoView({ behavior: 'smooth' });
    setIsOpen(false);
  };

  const scrollToBottom = () => {
    document.getElementById('end')?.scrollIntoView({ behavior: 'smooth' });
    setIsOpen(false);
  };

  return (
    <>
      <div className="relative z-20">
        <AnimatePresence>
          {isOpen && (
            <>
              {/* Top action */}
              <motion.button
                className="fab-action bg-green-500 text-white"
                style={{ bottom: '80px', right: '8px' }}
                initial={{ opacity: 0, y: 20 }}
                animate={{ opacity: 1, y: 0 }}
                exit={{ opacity: 0, y: 20 }}
                transition={{ duration: 0.2 }}
                onClick={scrollToTop}
                aria-label="Scroll to top"
              >
                <svg xmlns="http://www.w3.org/2000/svg" className="h-5 w-5" viewBox="0 0 20 20" fill="currentColor">
                  <path fillRule="evenodd" d="M14.707 12.707a1 1 0 01-1.414 0L10 9.414l-3.293 3.293a1 1 0 01-1.414-1.414l4-4a1 1 0 011.414 0l4 4a1 1 0 010 1.414z" clipRule="evenodd" />
                </svg>
              </motion.button>
              
              {/* Bottom action */}
              <motion.button
                className="fab-action bg-blue-500 text-white"
                style={{ bottom: '135px', right: '8px' }}
                initial={{ opacity: 0, y: 20 }}
                animate={{ opacity: 1, y: 0 }}
                exit={{ opacity: 0, y: 20 }}
                transition={{ duration: 0.2, delay: 0.05 }}
                onClick={scrollToBottom}
                aria-label="Scroll to bottom"
              >
                <svg xmlns="http://www.w3.org/2000/svg" className="h-5 w-5" viewBox="0 0 20 20" fill="currentColor">
                  <path fillRule="evenodd" d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z" clipRule="evenodd" />
                </svg>
              </motion.button>
              
              {/* Copy email action */}
              <CopyToClipboard text={email} onCopy={handleCopy}>
                <motion.button
                  className="fab-action bg-yellow-500 text-white"
                  style={{ bottom: '190px', right: '8px' }}
                  initial={{ opacity: 0, y: 20 }}
                  animate={{ opacity: 1, y: 0 }}
                  exit={{ opacity: 0, y: 20 }}
                  transition={{ duration: 0.2, delay: 0.1 }}
                  aria-label="Copy email address"
                >
                  <svg xmlns="http://www.w3.org/2000/svg" className="h-5 w-5" viewBox="0 0 20 20" fill="currentColor">
                    <path d="M8 3a1 1 0 011-1h2a1 1 0 110 2H9a1 1 0 01-1-1z" />
                    <path d="M6 3a2 2 0 00-2 2v11a2 2 0 002 2h8a2 2 0 002-2V5a2 2 0 00-2-2 3 3 0 01-3 3H9a3 3 0 01-3-3z" />
                  </svg>
                </motion.button>
              </CopyToClipboard>
            </>
          )}
        </AnimatePresence>

        {/* Main FAB button */}
        <motion.button
          className="fab bg-primary-600 text-white focus:outline-none focus:ring-2 focus:ring-primary-500 focus:ring-offset-2"
          onClick={() => setIsOpen(!isOpen)}
          whileHover={{ scale: 1.05 }}
          whileTap={{ scale: 0.95 }}
          aria-label={isOpen ? "Close menu" : "Open menu"}
          aria-expanded={isOpen}
        >
          {isOpen ? (
            <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M6 18L18 6M6 6l12 12" />
            </svg>
          ) : (
            <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M4 6h16M4 12h16m-7 6h7" />
            </svg>
          )}
        </motion.button>
      </div>

      {/* Toast notification */}
      <AnimatePresence>
        {showToast && (
          <motion.div
            className="toast"
            initial={{ opacity: 0, y: 50 }}
            animate={{ opacity: 1, y: 0 }}
            exit={{ opacity: 0 }}
            transition={{ duration: 0.3 }}
            role="alert"
          >
            Email copied to clipboard!
          </motion.div>
        )}
      </AnimatePresence>
    </>
  );
};

export default FloatingActionButton;
