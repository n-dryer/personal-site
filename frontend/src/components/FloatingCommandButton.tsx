import React, { useState, useEffect } from 'react';
import { motion, AnimatePresence, Variants } from 'framer-motion';
import { Command } from 'lucide-react';

interface FloatingCommandButtonProps {
  toggleCommandMenu: () => void;
}

export const FloatingCommandButton: React.FC<FloatingCommandButtonProps> = ({
  toggleCommandMenu,
}) => {
  const [showTooltip, setShowTooltip] = useState<boolean>(false);
  const [isMac, setIsMac] = useState<boolean>(false);
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [animationPhase, setAnimationPhase] = useState<'hidden' | 'visible' | 'pulse'>('hidden');
  const [tooltipTimer, setTooltipTimer] = useState<ReturnType<typeof setTimeout> | null>(null);

  // Detect OS for keyboard shortcut display
  useEffect(() => {
    setIsMac(navigator.platform.toUpperCase().indexOf('MAC') >= 0);
  }, []);

  // Synchronized delayed appearance with scroll button
  useEffect(() => {
    const delayTimer = setTimeout(() => {
      setIsVisible(true);
      setAnimationPhase('visible');

      // Start pulse after fade-in completes (0.75s delay + 1s fade-in = 1.75s total)
      setTimeout(() => {
        setAnimationPhase('pulse');
      }, 1000); // Start pulse after fade-in completes
    }, 750); // 0.75 second delay - synchronized with scroll button

    return () => clearTimeout(delayTimer);
  }, []);

  // Cleanup tooltip timer on unmount
  useEffect(() => {
    return () => {
      if (tooltipTimer) {
        clearTimeout(tooltipTimer);
      }
    };
  }, [tooltipTimer]);

  // Handle mouse enter with 500ms delay
  const handleMouseEnter = (): void => {
    // Clear any existing timer
    if (tooltipTimer) {
      clearTimeout(tooltipTimer);
    }

    // Set new timer for 500ms delay
    const timer = setTimeout(() => {
      setShowTooltip(true);
    }, 500);

    setTooltipTimer(timer);
  };

  // Handle mouse leave with immediate hide
  const handleMouseLeave = (): void => {
    // Clear any pending timer
    if (tooltipTimer) {
      clearTimeout(tooltipTimer);
      setTooltipTimer(null);
    }

    // Immediately hide tooltip
    setShowTooltip(false);
  };

  // Check for reduced motion preference safely
  const [prefersReducedMotion, setPrefersReducedMotion] = useState<boolean>(false);

  // Set up reduced motion detection
  useEffect(() => {
    if (typeof window !== 'undefined' && window.matchMedia) {
      const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
      setPrefersReducedMotion(mediaQuery.matches);

      const handleChange = (e: MediaQueryListEvent) => {
        setPrefersReducedMotion(e.matches);
      };

      mediaQuery.addEventListener('change', handleChange);
      return () => mediaQuery.removeEventListener('change', handleChange);
    }
  }, []);

  // Framer Motion variants for synchronized animations
  const buttonVariants: Variants = {
    visible: {
      opacity: 1,
      scale: 1,
      y: 0,
      transition: {
        opacity: { duration: 1, ease: 'easeOut' }, // 1 second fade-in duration - synchronized
        scale: {
          type: 'spring',
          stiffness: 300,
          damping: 20,
          duration: 0.5
        },
        y: {
          type: 'spring',
          stiffness: 300,
          damping: 20,
          duration: 1 // 1 second upward slide - synchronized
        }
      }
    },
    hidden: {
      opacity: 0,
      scale: 0.8,
      y: 16, // Synchronized initial position (translate-y-4 = 16px)
      transition: {
        duration: 0.2,
        ease: 'easeIn'
      }
    },
    pulse: {
      scale: [1, 1.02, 1], // Gentle pulse animation
      transition: {
        duration: 3,
        repeat: Infinity,
        repeatType: 'loop' as const,
        ease: 'easeInOut'
      }
    },
    hover: {
      scale: 1.05,
      boxShadow: '0 10px 25px -5px rgba(59, 130, 246, 0.4)',
      transition: { duration: 0.2 }
    },
    tap: {
      scale: 0.95,
      transition: { duration: 0.1 }
    }
  };

  return (
    <AnimatePresence>
      {isVisible && (
        <motion.div
          className="fixed bottom-6 right-6 z-40"
          initial="hidden"
          animate="visible"
          exit="hidden"
          variants={buttonVariants}
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
        >
          <AnimatePresence>
            {showTooltip && (
              <motion.div
                initial={{ opacity: 0, scale: 0.9, y: 10 }}
                animate={{ opacity: 1, scale: 1, y: 0 }}
                exit={{ opacity: 0, scale: 0.9, y: 10 }}
                transition={{ duration: 0.2 }}
                className="absolute bottom-full right-0 mb-4 whitespace-nowrap rounded-lg px-4 py-2 text-sm text-white shadow-lg backdrop-blur-md"
                style={{ background: 'var(--token-bg-frosted)' }}
              /* Frosted glass effect for overlays */
              >
                <div className="flex items-center gap-2">
                  <span className="font-medium">Open Command Menu</span>
                  <span className="text-xs text-white/70">
                    {isMac ? '⌘K' : 'Ctrl+K'}
                  </span>
                </div>
                <div className="absolute bottom-0 right-4 h-2 w-2 translate-y-1/2 rotate-45 transform bg-white/10"></div>
              </motion.div>
            )}
          </AnimatePresence>

          <motion.button
            onClick={toggleCommandMenu}
            className="flex h-14 w-14 items-center justify-center rounded-full text-white shadow-lg transition-all duration-300 bg-gradient-button hover:bg-gradient-button-hover focus:outline-none focus-visible:ring-2 focus-visible:ring-token-primary-500 focus-visible:ring-offset-2 dark:focus-visible:ring-offset-token-secondary-900"
            whileHover="hover"
            whileTap="tap"
            variants={buttonVariants}
            animate={prefersReducedMotion ? undefined : animationPhase} // Controlled animation phases
            aria-label="Open Command Menu"
            aria-keyshortcuts={isMac ? '⌘+K' : 'Ctrl+K'}
            role="button"
            tabIndex={0}
            onKeyDown={(e) => {
              if (e.key === 'Enter' || e.key === ' ') {
                e.preventDefault();
                toggleCommandMenu();
              }
            }}
          >
            <Command size={24} />
          </motion.button>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
