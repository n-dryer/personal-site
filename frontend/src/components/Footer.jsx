import React from 'react';
import { motion } from 'framer-motion';
import { SocialIcon } from 'react-social-icons';

const Footer = ({ userData }) => {
  const currentYear = new Date().getFullYear();

  return (
    <footer id="end" className="bg-gray-900 text-white py-12">
      <div className="container mx-auto px-4">
        <div className="flex flex-col md:flex-row justify-between items-center">
          <div className="mb-6 md:mb-0">
            <h2 className="text-2xl font-bold mb-2">{userData.fullName}</h2>
            <p className="text-gray-400">{userData.bioLine}</p>
          </div>
          
          <div className="flex space-x-4">
            {userData.socialLinks.map((link, index) => (
              <motion.div
                key={index}
                whileHover={{ scale: 1.1, rotate: 5 }}
                whileTap={{ scale: 0.9 }}
              >
                <SocialIcon 
                  url={link.url}
                  className="transition-transform duration-300"
                  target="_blank"
                  rel="noopener noreferrer"
                  style={{ height: 40, width: 40 }}
                  aria-label={`Visit ${link.name}`}
                />
              </motion.div>
            ))}
          </div>
        </div>
        
        <div className="border-t border-gray-800 mt-8 pt-8 flex flex-col md:flex-row justify-between items-center">
          <p className="text-gray-400 mb-4 md:mb-0">
            &copy; {currentYear} {userData.fullName}. All rights reserved.
          </p>
          <div className="text-gray-400">
            <a 
              href={`mailto:${userData.email}`}
              className="hover:text-primary-500 transition-colors duration-300"
            >
              {userData.email}
            </a>
          </div>
        </div>
      </div>
    </footer>
  );
};

export default Footer;
