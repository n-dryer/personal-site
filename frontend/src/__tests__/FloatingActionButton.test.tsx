import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { FloatingActionButton } from '../components/FloatingActionButton';

// Mock framer-motion
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: React.PropsWithChildren<any>) => (
      <div data-testid="motion-div" {...props}>{children}</div>
    ),
    button: ({ children, ...props }: React.PropsWithChildren<any>) => (
      <button data-testid="motion-button" {...props}>{children}</button>
    ),
  },
  AnimatePresence: ({ children }: React.PropsWithChildren<any>) => <>{children}</>,
}));

// Mock lucide-react
jest.mock('lucide-react', () => ({
  Command: () => <div data-testid="command-icon">Command Icon</div>,
}));

describe('FloatingActionButton Component', () => {
  // Mock props
  const mockToggleCommandMenu = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
    
    // Mock navigator.platform for OS detection
    Object.defineProperty(navigator, 'platform', {
      configurable: true,
      value: 'MacIntel',
    });
    
    // Reset timers before each test
    jest.useRealTimers();
  });
  
  it('renders correctly', () => {
    const { container } = render(
      <FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />
    );
    expect(container).toMatchSnapshot();
  });
  
  it('shows tooltip initially', () => {
    render(<FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />);
    
    // Tooltip should be visible with Mac keyboard shortcut
    expect(screen.getByText(/press/i)).toBeInTheDocument();
    expect(screen.getByText('⌘+K')).toBeInTheDocument();
  });
  
  it('hides tooltip after timeout', () => {
    jest.useFakeTimers();
    
    render(<FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />);
    
    // Tooltip should be visible initially
    expect(screen.getByText(/press/i)).toBeInTheDocument();
    
    // Fast-forward timers to trigger tooltip hide
    jest.advanceTimersByTime(5000);
    
    // Tooltip should be hidden
    expect(screen.queryByText(/press/i)).not.toBeInTheDocument();
  });
  
  it('shows Windows shortcut when not on Mac', () => {
    // Mock Windows platform
    Object.defineProperty(navigator, 'platform', {
      configurable: true,
      value: 'Win32',
    });
    
    render(<FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />);
    
    // Should show Ctrl+K instead of ⌘+K
    expect(screen.getByText('Ctrl+K')).toBeInTheDocument();
  });
  
  it('calls toggleCommandMenu when button is clicked', () => {
    render(<FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />);
    
    // Find the button and click it
    const button = screen.getByTestId('motion-button');
    fireEvent.click(button);
    
    // Should call the toggle function
    expect(mockToggleCommandMenu).toHaveBeenCalledTimes(1);
  });
  
  it('shows tooltip on mouse enter after it was hidden', () => {
    jest.useFakeTimers();
    
    render(<FloatingActionButton toggleCommandMenu={mockToggleCommandMenu} />);
    
    // Hide tooltip first
    jest.advanceTimersByTime(5000);
    expect(screen.queryByText(/press/i)).not.toBeInTheDocument();
    
    // Trigger mouse enter
    const container = screen.getByTestId('motion-button').closest('div') as HTMLElement;
    fireEvent.mouseEnter(container);
    
    // Tooltip should be visible again
    expect(screen.getByText(/press/i)).toBeInTheDocument();
  });
});
