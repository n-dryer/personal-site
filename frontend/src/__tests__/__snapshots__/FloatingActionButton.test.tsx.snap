// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingActionButton Component renders correctly 1`] = `
<div>
  <div
    class="fixed bottom-6 right-6 z-40"
  >
    <div
      animate="[object Object]"
      class="absolute bottom-full mb-4 right-0 bg-white dark:bg-gray-800 text-gray-700 dark:text-gray-200 px-3 py-2 rounded-lg shadow-md text-sm"
      data-testid="motion-div"
      exit="[object Object]"
      initial="[object Object]"
    >
      Press 
      <kbd
        class="px-1.5 py-0.5 text-xs rounded bg-gray-100 dark:bg-gray-700 text-gray-700 dark:text-gray-300 font-mono"
      >
        âŒ˜
        +K
      </kbd>
       for commands
    </div>
    <button
      aria-label="Open command palette"
      class="w-14 h-14 rounded-full shadow-lg flex items-center justify-center focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2 dark:focus-visible:ring-offset-gray-900 bg-blue-500 text-white hover:bg-blue-600 transition-colors duration-300"
      data-testid="motion-button"
      whilehover="[object Object]"
      whiletap="[object Object]"
    >
      <div
        data-testid="command-icon"
      >
        Command Icon
      </div>
    </button>
  </div>
</div>
`;
